### PR TITLE
Use ACID semantics for whole partition deletes

### LINKED ISSUE #7291: Non-transactional DELETE from ORC ACID when predicate on partition columns
Setup

```
trino:default> CREATE TABLE nation WITH (transactional=true, partitioned_by=ARRAY['regionkey']) AS SELECT nationkey, regionkey FROM tpch.tiny.nation;
CREATE TABLE: 25 rows
```

### Delete via Hive

```
0: jdbc:hive2://localhost:10000/default> DELETE FROM nation WHERE regionkey = 4;
```

```
$ hdfs dfs -ls -R /user/hive/warehouse/nation
drwxrwxrwx   - hive supergroup          0 2021-03-15 17:01 /user/hive/warehouse/nation/regionkey=0
-rw-r--r--   3 hive supergroup        275 2021-03-15 17:01 /user/hive/warehouse/nation/regionkey=0/000000_128931710732837179949886632921173018761
drwxrwxrwx   - hive supergroup          0 2021-03-15 17:01 /user/hive/warehouse/nation/regionkey=1
-rw-r--r--   3 hive supergroup        275 2021-03-15 17:01 /user/hive/warehouse/nation/regionkey=1/000000_124847042386740602256668008681862351404
drwxrwxrwx   - hive supergroup          0 2021-03-15 17:01 /user/hive/warehouse/nation/regionkey=2
-rw-r--r--   3 hive supergroup        278 2021-03-15 17:01 /user/hive/warehouse/nation/regionkey=2/000000_1216022056472270887212432062596750395615
drwxrwxrwx   - hive supergroup          0 2021-03-15 17:01 /user/hive/warehouse/nation/regionkey=3
-rw-r--r--   3 hive supergroup        278 2021-03-15 17:01 /user/hive/warehouse/nation/regionkey=3/000000_138121640554415214739608071094296332622
drwxrwxrwx   - hive supergroup          0 2021-03-15 17:02 /user/hive/warehouse/nation/regionkey=4
-rw-r--r--   3 hive supergroup        275 2021-03-15 17:01 /user/hive/warehouse/nation/regionkey=4/000000_1235534928977361567916286797118886462565
drwxr-xr-x   - hive supergroup          0 2021-03-15 17:02 /user/hive/warehouse/nation/regionkey=4/delete_delta_0000001_0000001_0000
-rw-r--r--   3 hive supergroup          1 2021-03-15 17:02 /user/hive/warehouse/nation/regionkey=4/delete_delta_0000001_0000001_0000/_orc_acid_version
-rw-r--r--   3 hive supergroup        654 2021-03-15 17:02 /user/hive/warehouse/nation/regionkey=4/delete_delta_0000001_0000001_0000/bucket_00000
```

### Delete via Trino

Recreate data

```
trino:default> DROP TABLE nation;
DROP TABLE
trino:default> CREATE TABLE nation WITH (transactional=true, partitioned_by=ARRAY['regionkey']) AS SELECT nationkey, regionkey FROM tpch.tiny.nation;
CREATE TABLE: 25 rows
```

```
trino:default> DELETE FROM nation WHERE regionkey = 4;
DELETE
```

```
$ hdfs dfs -ls -R /user/hive/warehouse/nation
drwxrwxrwx   - hive supergroup          0 2021-03-15 17:04 /user/hive/warehouse/nation/regionkey=0
-rw-r--r--   3 hive supergroup        274 2021-03-15 17:04 /user/hive/warehouse/nation/regionkey=0/000000_1063124892134058974316784284952306601345
drwxrwxrwx   - hive supergroup          0 2021-03-15 17:04 /user/hive/warehouse/nation/regionkey=1
-rw-r--r--   3 hive supergroup        274 2021-03-15 17:04 /user/hive/warehouse/nation/regionkey=1/000000_991844335898659669411016087017478636796
drwxrwxrwx   - hive supergroup          0 2021-03-15 17:04 /user/hive/warehouse/nation/regionkey=2
-rw-r--r--   3 hive supergroup        277 2021-03-15 17:04 /user/hive/warehouse/nation/regionkey=2/000000_1286352826480616106215703064477961637337
drwxrwxrwx   - hive supergroup          0 2021-03-15 17:04 /user/hive/warehouse/nation/regionkey=3
-rw-r--r--   3 hive supergroup        277 2021-03-15 17:04 /user/hive/warehouse/nation/regionkey=3/000000_107610126727478156749651387731393069546
```

### Observed

Whole partition directory deleted.

### Expected

Delete from a transactional table should be transactional and follow Hive behavior for compatibility.
Thus  `delete_delta_*` should be created instead of deleting the original data files.

### PR DESCRIPTION
**Pull request by electrum**
Use a row level delete instead of dropping partitions when running a delete that exactly matches entire partitions.

Fixes #7291

### CODE REVIEW

File: testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestHiveTransactionalTable.java [hunk 1]
```diff
@@ -1283,10 +1283,21 @@ public void testDeleteWholePartition()
         withTemporaryTable("delete_partitioned", true, true, NONE, tableName -> {
             onTrino().executeQuery(format("CREATE TABLE %s WITH (transactional = true, partitioned_by = ARRAY['regionkey'])" +
                     " AS SELECT nationkey, name, regionkey FROM tpch.tiny.nation", tableName));
+
             // SELECT before deletion may prime the cache on the Hive side
             verifySelectForPrestoAndHive("SELECT count(*) FROM " + tableName, "true", row(25));
+
+            // verify all partitions exist
+            assertThat(onTrino().executeQuery(format("SELECT * FROM \"%s$partitions\"", tableName)))
```
Comment (djsstarburst) [inline testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestHiveTransactionalTable.java:1291]: Nice!

**Review by djsstarburst — APPROVED** [review]
LGTM

**Review by losipiuk — APPROVED** [review]
LGTM. Thank you

**Comment by sopel39** [comment]
Curious, does it have performance impact vs bulk remove of partition?

**Comment by electrum** [comment]
@sopel39 yes, the old way was basically constant time, since it just dropped partitions, whereas this is proportional to the number of rows deleted. That's why we didn't do it this way originally. (it was an explicit choice that we discussed, not a bug or oversight)

**Pull request merged**